Basic validation against certificate authority

lambrojos opened the pull request:
Hello, 
I needed to verify a certificate against a chain of CA certificates, I adapted code from http://fm4dd.com/openssl/certverify.htm, and now I am trying to contribute back, I think that verification of a certificate is an operation that goes well alongside parsing.

I also threw  in some basic typings for typescript.. should I have done another PR?

yorkie commented on index.js:
```diff
@@ -1,10 +1,19 @@
+
+exports.verify = function(certPath, CABundlePath){
+  assert(certPath, 'certificate path is required');
+  assert(CABundlePath, 'certificate path is required');
```
Why do you use assert to instead of throwing errors?

yorkie commented on index.js:
```diff
@@ -1,10 +1,19 @@
+  assert(certPath, 'certificate path is required');
+  assert(CABundlePath, 'certificate path is required');
+  return x509.verify(certPath, CABundlePath);
+}
```
semi-colon

yorkie commented on node-x509.d.ts:
```diff
@@ -0,0 +1,9 @@
+  function verify(certPath: string, CACertChainPath: string): any
+}
+
+export = x509;
```
Could you please add this in another pull-request, this is not relevant about the subject of this pull-request.

yorkie commented on package.json:
```diff
@@ -11,8 +11,10 @@
   "scripts": {
     "test": "node test/test"
   },
+  "typings": "./node-x509.d.ts",
```
Ditto

yorkie commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "x509",
-  "version": "0.2.6",
+  "version": "0.3.0",
```
Please do not change the version in any specific feature pull-request, thanks.

yorkie commented on package.json:
```diff
@@ -11,8 +11,10 @@
   "dependencies": {
-    "nan": "2.2.0"
+    "nan": "2.2.0",
+    "node-gyp": "^3.4.0"
```
Use precise version string, thank you :-)

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+  std::string ca_bundlestr = *String::Utf8Value(info[1]->ToString());
+  /*
+    Create x509 cert
+   */
```
use // to comment here :-)

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+    Create x509 cert
+   */
+  X509_STORE *store = NULL;
+  X509_STORE_CTX *vrfy_ctx = NULL;
```
`verify_ctx` please :-)

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+  X509_STORE *store = NULL;
+  X509_STORE_CTX *vrfy_ctx = NULL;
+  X509 *cert = NULL;
+  BIO *certbio = BIO_new(BIO_s_file());
```
`cert_bio`

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+  X509 *cert = NULL;
+  BIO *certbio = BIO_new(BIO_s_file());
+
+  if (!(store=X509_STORE_new())){
```
whitespaces, thank you :-)

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+    Nan::ThrowError("Failed to create X509 certificate store.");
+  }
+
+  vrfy_ctx = X509_STORE_CTX_new();
```
Should we check `verify_ctx` is `NULL` as here the OpenSSL noted:

> X509_STORE_CTX_new() returns an newly allocates context or NULL is an error occurred.

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+  vrfy_ctx = X509_STORE_CTX_new();
+
+  // load file in BIO
+  int ret = BIO_read_filename(certbio, cert_path.c_str());
```
Check if `BIO_read_filename` has been operated successfully :-)

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,67 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+  // load file in BIO
+  int ret = BIO_read_filename(certbio, cert_path.c_str());
+
+  if (! (cert = PEM_read_bio_X509(certbio, NULL, 0, NULL))) {
```
Whitespace :-)

yorkie replied:
By the way, could you please separate this complex statement to clear subsequent ones like this:

```cpp
cert = PEM_read_bio_X509(cert_bio, NULL, 0, NULL);
if (!cert) {
  // ...
}
```

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,16 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+  path.join(__dirname,'certs/enduser-example.com.crt'),
+  path.join(__dirname,'CA_chains/enduser-example.com.chain')
+)
+assert(verified)
```
Could you please add a scoping brackets here to avoid the variable `verified` accessible :-)

yorkie commented:
@lambrojos thanks for this feature pull request, and leave some comments/nits, and before merging your changes into master, you have to fix them firstly :-)

lambrojos commented:
Thanks for the review and the feedback !  I will fix as requested :)
BTW, should I add something to the documentation?

yorkie commented:
> BTW, should I add something to the documentation?

I think so, @lambrojos thank you :-)

lambrojos commented:
Hi, I hope I have addressed all your comments :) I have also added a stat check on argument files, this is because a wrong path resulted in a segmentation fault

Southern commented on test/test.js:
```diff
@@ -1,12 +1,37 @@
 fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
   console.log("File: %s", file);
-  console.log(x509.parseCert(path.join(__dirname, 'certs', file)));
+  //console.log(x509.parseCert(path.join(__dirname, 'certs', file)));
```
These tests still need to run.

yorkie commented on index.js:
```diff
@@ -1,11 +1,25 @@
+  fs.statSync(certPath);
+  fs.statSync(CABundlePath);
+  return x509.verify(certPath, CABundlePath);
+};
```
This function is not follow the coding style :(

yorkie commented:
@lambrojos please take a look at other parts in this project, and learn what's the coding style, thank you :)

lambrojos commented:
Ok, I hope I got the coding style right this time...

yorkie commented on index.js:
```diff
@@ -1,11 +1,24 @@
 exports.getSubject = x509.getSubject;
 exports.getIssuer = x509.getIssuer;
 
+exports.verify = function(certPath, CABundlePath){
```
Should have white space after the brace.

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,84 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+
+  // create store
+  store = X509_STORE_new();
+  if (!store){
```
Ditto

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,84 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+
+  verify_ctx = X509_STORE_CTX_new();
+
+  if (verify_ctx == NULL){
```
Coding style..

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,84 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+
+  // load file in BIO
+  int ret = BIO_read_filename(cert_bio, cert_path.c_str());
+  if (ret != 1){
```
Ditto...

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,84 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+
+  // load CA bundle
+  ret = X509_STORE_load_locations(store, ca_bundlestr.c_str(), NULL);
+  if (ret != 1){
```
Ditto...

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+  var verified = x509.verify(
+    path.join(__dirname,'certs/enduser-example.com.crt'),
+    path.join(__dirname,'CA_chains/enduser-example.com.chain')
+  )
```
semi-colon is missing..

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+    path.join(__dirname,'certs/enduser-example.com.crt'),
+    path.join(__dirname,'CA_chains/enduser-example.com.chain')
+  )
+  assert(verified)
```
Missing semi-colon...

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+    path.join(__dirname,'CA_chains/enduser-example.com.chain')
+  )
+  assert(verified)
+})
```
Ditto

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
   console.log();
 });
+
+(function(){
```
Use braces instead the unneeded function wrap :)

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
 });
+
+(function(){
+  var verified = x509.verify(
```
With brackets scoping, we have to use `const` here.

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+})
+
+
+assert.throws(function(){
```
White space is missing.

yorkie commented on test/test.js:
```diff
@@ -10,3 +11,27 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+}, /unable to get local issuer certificate/)
+
+
+assert.throws(function(){
```
White space.

yorkie commented on README.md:
```diff
@@ -112,6 +112,29 @@ cert = { subject:
+
+x509.verify(
+  __dirname + '/certs/user.com.crt',
+  __dirname + 'enduser-example.com.chain',
```
Please remove the colon of this line that is unneeded.

yorkie commented:
@lambrojos Thanks for your quick addressing, but it seems to have few nits about coding style and others, I have listed them and may I ask to fix them again, thank you :)

yorkie commented on test/test.js:
```diff
@@ -1,6 +1,7 @@
 var x509 = require('../index'),
     fs = require('fs'),
-    path = require('path');
+    path = require('path')
```
Missing a colon

yorkie commented on binding.gyp:
```diff
@@ -48,7 +48,7 @@
+      ],
     }
   ]
 }
```
The change on this file is not needed, please remove :)

yorkie commented on index.js:
```diff
@@ -1,11 +1,24 @@
+    throw new Error('CA Bundle path is required');
+  }
+  fs.statSync(certPath);
+  fs.statSync(CABundlePath);
```
Could we please provide the async function?

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,85 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+
+NAN_METHOD(verify) {
+  Nan::HandleScope scope;
+  OpenSSL_add_all_algorithms();
```
Node.js has added algorithms at [here](https://github.com/nodejs/node/blob/master/src/node_crypto.cc#L5778), so please remove this line.

lambrojos replied:
That is true, but I suspect the code you pointed at only runs if the crypto module is required..

In fact if I remove the line the test fails with
 ```
"Error: certificate signature failure".
```
I am quite sure that the certificate signature is correct, because if I run openssl verify I get the expected result:

```
:~/code/node-x509$ openssl verify -CAfile test/CA_chains/enduser-example.com.chain test/certs/enduser-example.com.crt 
test/certs/enduser-example.com.crt: OK
```

If I add `require('crypto')` before the `verify()` function gets called,  everything is fine again.

(I obtained this result on node 4.4.3 and 6.2.2)

So I believe we have to choose to either require the crypto module in index.js or to keep that line. Maybe requiring crypto is safer? What do you suggest?

yorkie replied:
Requiring `crypto` looks better to me :)

Southern replied:
Not sure that pulling in another module is the best solution, especially when we're not using anything in the crypto module other than it loading the algorithms. Calling `OpenSSL_add_all_algorithms` in our code won't cause problems if someone uses `require('crypto')`, as subsequent calls to `OpenSSL_add_all_algorithms` will work perfectly fine.

lambrojos replied:
That's right, on my dev machine I have an earlier version:
```
openssl version
OpenSSL 1.0.2d 9 Jul 2015
```
So, what are we going to do? should I leave the line as it is?

lambrojos replied:
Or maybe use a preprocessor directive?
```c
  # if OPENSSL_API_COMPAT < 0x10100000L
    OpenSSL_add_all_algorithms();
  # endif
```

Southern replied:
I would say to leave that line out entirely and note in the documentation that it may only work on OpenSSL 1.1.0 or greater. Most people should have already upgraded to 1.1.0, and those who haven't should, due to the heartbleed vulnerability that was discovered.

Southern replied:
In fact, they should be able to `require('crypto')` in their own code if they want to use it in a version earlier than 1.1.0 since it calls `OpenSSL_add_all_algorithms` when it's loaded. That would allow them to remove the dependency on their own when they upgrade without us having to change our code.

lambrojos replied:
Ok, I'll remove the line and add a warning in the docs

yorkie replied:
+1 on https://github.com/Southern/node-x509/pull/38#discussion_r78301382 :)

Southern replied:
@yorkie you think the preprocessor directive would be better?

Southern replied:
@lambrojos May want to make sure that I'm right about `require('crypto')` working the way that I'm thinking it will before you go through the trouble. It may be that we need the preprocessor directive. I'm just hoping that we don't.

yorkie replied:
Actually I'm not sure which would be better, Node.js core doesn't give the best solution for this. Now the node-x509 and the core module `crypto` are in parallel to use OpenSSL. Loading `crypto` at user-land would not be user-friend in some ways?

Calling `OpenSSL_add_all_algorithms()` without directives looks good to me as well now :)

Southern replied:
Yeah, I was wondering about the user-friendliness of it. I'm good with
whichever since either of them will solve the problem. Just make sure that
the version is bumped to 0.3.0 when this is merged.


From: Yorkie Liu <notifications@github.com> <notifications@github.com>
Reply: Southern/node-x509
<reply@reply.github.com>
<reply@reply.github.com>
Date: September 12, 2016 at 1:19:42 PM
To: Southern/node-x509 <node-x509@noreply.github.com>
<node-x509@noreply.github.com>
Cc: Colton Baker <educability@gmail.com> <educability@gmail.com>, Comment
<comment@noreply.github.com> <comment@noreply.github.com>
Subject:  Re: [Southern/node-x509] Basic validation against certificate
authority (#38)

In src/x509.cc
<https://github.com/Southern/node-x509/pull/38#discussion_r78414738>:

> @@ -45,6 +45,85 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
>    return *String::Utf8Value(info[0]->ToString());
>  }
>
> +
> +
> +NAN_METHOD(verify) {
> +  Nan::HandleScope scope;
> +  OpenSSL_add_all_algorithms();

Actually I'm not sure which would be better, Node.js core doesn't give the
best solution for this. Now the node-x509 and the core module crypto are in
parallel to use OpenSSL. Loading crypto at user-land would not be
user-friend in some ways?

Calling OpenSSL_add_all_algorithms() without directives looks good to me as
well now :)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/Southern/node-x509/pull/38/files/0353b2b4e9d464638a5c7e1a2c264d559a483bdb#r78414738>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AA2JXfXzd3sDuxQlXUx8CQvwmDuANunDks5qpYmugaJpZM4Jmuix>
.

yorkie replied:
So I think now it's up to you @lambrojos :)

lambrojos replied:
ook, so I'll leave the line as it is, no directives
.. right?

yorkie replied:
Yea :)

yorkie commented on src/x509.cc:
```diff
@@ -45,6 +45,85 @@ std::string parse_args(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+
+  // create store
+  store = X509_STORE_new();
+  if (!store) {
```
use `store === NULL`

yorkie commented on test/test.js:
```diff
@@ -13,26 +13,30 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+  path.join(__dirname, 'certs/enduser-example.com.crt'),
+  path.join(__dirname, 'CA_chains/enduser-example.com.chain'),
+  function (err) {
+    assert(err === null);
```
use `assert.strictEqual`

yorkie commented on test/test.js:
```diff
@@ -13,26 +13,30 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+  path.join(__dirname, 'CA_chains/enduser-example.com.chain'),
+  function (err, result) {
+    assert(err instanceof Error);
+    assert(err.message === 'unable to get local issuer certificate');
```
Ditto

yorkie commented on test/test.js:
```diff
@@ -13,26 +13,30 @@ fs.readdirSync(path.join(__dirname, 'certs')).forEach(function (file) {
+    assert(err instanceof Error);
+    assert(err.message.match(/ENOENT: no such file or directory/));
+  }
+);
```
Try to use `assert.throws`

yorkie commented on index.js:
```diff
@@ -6,16 +6,35 @@ exports.getAltNames = x509.getAltNames;
-exports.verify = function(certPath, CABundlePath) {
+exports.verify = function(certPath, CABundlePath, cb) {
   if (!certPath) {
     throw new Error('Certificate path is required');
```
TypeError please :)

yorkie commented on index.js:
```diff
@@ -6,16 +6,35 @@ exports.getAltNames = x509.getAltNames;
     throw new Error('Certificate path is required');
   }
   if (!CABundlePath) {
     throw new Error('CA Bundle path is required');
```
Ditto: TypeError

yorkie commented on index.js:
```diff
@@ -6,16 +6,35 @@ exports.getAltNames = x509.getAltNames;
+      catch (verificationError) {
+        cb(verificationError);
+      }
+    })
```
Semi-colon :)

yorkie commented:
LGTM if @Southern has no objection, I will land this tonight, thank you @lambrojos 👍

yorkie commented:
Bumped v0.3.0, thank you @lambrojos @Southern :)

lambrojos commented:
Yaay thanks guys!
